compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -325,6 +328,100 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
+        &mut self,
+        restriction: &'ast ast::Restriction<Kind>,
+        finalize: bool,
+    ) -> Result<ty::Restriction, RestrictionResolutionError<'ast>> {
+        let parent_scope = &self.parent_scope;
+        match restriction.level {
+            ast::RestrictionLevel::Unrestricted => Ok(ty::Restriction::Unrestricted),
+            // FIXME(jhpratt) create an "implied" level in rustc_middle. This will be useful for
+            // lints in addition to conversion to visibility.
+            // If the restriction is implied, it has no effect when the item is otherwise visible.
+            ast::RestrictionLevel::Implied => Ok(ty::Restriction::Unrestricted),
+            ast::RestrictionLevel::Restricted { ref path, id, shorthand: _ } => {
+                // For restrictions we are not ready to provide correct implementation of "uniform
+                // paths" right now, so on 2018 edition we only allow module-relative paths for now.
+                // On 2015 edition visibilities are resolved as crate-relative by default,
+                // so we are prepending a root segment if necessary.
+                let ident = path.segments.get(0).expect("empty path in restriction").ident;
+                let crate_root = if ident.is_path_segment_keyword() {
+                    None
+                } else if ident.span.rust_2015() {
+                    Some(Segment::from_ident(Ident::new(
+                        kw::PathRoot,
+                        path.span.shrink_to_lo().with_ctxt(ident.span.ctxt()),
+                    )))
+                } else {
+                    return Err(RestrictionResolutionError::Relative2018(ident.span, path));
+                };
+
+                let segments = crate_root
+                    .into_iter()
+                    .chain(path.segments.iter().map(|seg| seg.into()))
+                    .collect::<Vec<_>>();
+                let expected_found_error = |res| {
+                    Err(RestrictionResolutionError::ExpectedFound(
+                        path.span,
+                        Segment::names_to_string(&segments),
+                        res,
+                    ))
+                };
+                match self.r.resolve_path(
+                    &segments,
+                    Some(TypeNS),
+                    parent_scope,
+                    finalize.then(|| Finalize::new(id, path.span)),
+                    None,
+                ) {
+                    PathResult::Module(ModuleOrUniformRoot::Module(module)) => {
+                        let res = module.res().expect("restriction resolved to unnamed block");
+                        if finalize {
+                            self.r.record_partial_res(id, PartialRes::new(res));
+                        }
+                        if module.is_normal() {
+                            if res == Res::Err {
+                                Ok(ty::Restriction::Unrestricted)
```

Comment:
Is the idea here that we had an error but we go ahead with Unrestricted to allow us to report other errors too without giving up immediately?

Reply:
Pretty much. It's the same as [this line](https://github.com/rust-lang/rust/blob/b22c152958eade17a71d899b29a2d39bcc77aa48/compiler/rustc_resolve/src/build_reduced_graph.rs#L302) for visibility continuing as if it's public.